lib/Doctrine/DBAL/SQLParserUtils.php

```diff
@@ -73,6 +73,42 @@ static public function getPlaceholderPositions($statement, $isPositional = true)
         return $paramMap;
     }
 
+    static private function startsWith($haystack, $needle)
+    {
+        $length = strlen($needle);
+        return (substr($haystack, 0, $length) === $needle);
+    }
+
+    static private function endsWith($haystack, $needle)
+    {
+        $length = strlen($needle);
+        if ($length == 0) {
+            return true;
+        }
+
+        return (substr($haystack, -$length) === $needle);
+    }
+
+    static private function isCollectionType($type)
+    {
+        if ($type === Connection::PARAM_INT_ARRAY || $type === Connection::PARAM_STR_ARRAY)
+            return true;
+
+        return static::startsWith($type, '[') && static::endsWith($type, ']');
+    }
+
+    static private function removeCollectionMarker($type)
+    {
+        if (!static::isCollectionType($type))
+            throw SQLParserUtilsException::notCollectionType($type);
+        
+        if (is_numeric($type))
+            return $type - Connection::ARRAY_PARAM_OFFSET;
+
+        // [my_custom_type] -> my_custom_type
```

Comment:
Ah, this is the trick: you are trying to use `[foo]` as an array mapping type of instances of `foo`.

The idea is good, but using strings for this is terrible, and we shouldn't let this land, as that is going to just complicate things much further.

A different approach could be needed here to define the type of a parameter: parsing the type is not viable...